src/Components/Components/src/Routing/Router.cs

```diff
@@ -227,7 +228,24 @@ private async ValueTask<bool> RunOnNavigateAsync(string path, Task previousOnNav
             }
 
             var completedTask = await Task.WhenAny(task, cancellationTcs.Task);
-            return task == completedTask;
+
+            if (completedTask.IsFaulted)
+            {
+                ThrowExceptionInRenderer(completedTask);
+                return false;
+            }
+
+            return task == completedTask && !(completedTask.IsCanceled || _onNavigateCts.Token.IsCancellationRequested);
+        }
+
+        internal virtual void ThrowExceptionInRenderer(Task task)
+        {
+            var exception = task.Exception.Flatten().InnerException;
+            while (exception is AggregateException e)
+            {
+                exception = e.InnerException;
+            }
```

Comment:
Do we need to be concerned with the case that the `AggregateException` has more than one [`InnerExceptions`](https://docs.microsoft.com/en-us/dotnet/api/system.aggregateexception.innerexceptions?view=netcore-3.1#System_AggregateException_InnerExceptions)?

I'd be totally fine with just dispatching the original `task.Exception` as-is and letting the upstream code do whatever it wants with reporting it sensibly. If that's not adequate would you mind clarifying why?

Reply:
We could throw the AggregateException back to the user and let them handle it but I'm biased towards unwrapped it ourselves because:

- It's not a fairly complicated task and we do it elsewhere in our codebase
- It means that the user doesn't have to understand the internals of our exception handling (it returns an aggregate exception from a task) and can just operate on the most meaningful piece of information (the inner exception)

Reply:
OK, fine with me! Thanks for clarifying. I was just unsure about whether there might be more complicated cases (multiple sibling inner exceptions) but I guess that's probably not applicable here.